plutus-tx/src/PlutusTx/Foldable.hs

```diff
@@ -225,6 +228,20 @@ find p = coerce . foldMap (\ x -> First (if p x then Just x else Nothing))
 -- | Plutus Tx version of 'Data.Foldable.mapM_'.
 {-# INLINABLE mapM_ #-}
 mapM_ :: (Foldable t, Haskell.Monad m) => (a -> m b) -> t a -> m ()
+-- See Note [Haskell.Monad]
 mapM_ f = foldr c (Haskell.return ())
   where c x k = f x Haskell.>> k
         {-# INLINE c #-}
+
+{- Note [Haskell.Monad]
+There is no `PlutusTx.Monad`. Since do-notations use `Haskell.Monad` (without
+`RebindableSyntax` anyway), it won't be very useful to have `PlutusTx.Monad`.
+
+`Haskell.Monad` instances for `Maybe`, `Either` and `[]` (and likely many other types)
+all work fine since the unfoldings of their `>>=` methods don't contain anything the
+plugin can't handle. This is of course not true in general, e.g., the plugin won't be
+able to handle a (hypothetical) `Haskell.Monad` instance whose `>>=` definition contains
+`GHC.Num.(+)` (until we make the plugin support functions like `GHC.Num.(+)`) or is
+not inlinable.
```

Comment:
This doesn't make any sense to me. We have this entirely separate hierarchy of type classes and then stop short of `Monad`, because of some concerns regarding `do`, which are easily resolvable with [`QualifiedDo`](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/qualified_do.html). I'm OK with using Haskell's type classes when they're magical like `Coercible`, but having our own `Applicative` and using Haskell's `Monad` is just extra weird.

Reply:
> which are easily resolvable with [QualifiedDo](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/qualified_do.html)

This is technically easy but I find it user-unfriendly, having to use different `do`s in different places. Going forward we should start reducing our own typeclass hierarchy rather than adding to it (which is what PLT-847 is about). Our `Applicative` and a few others can probably go away right now without affecting anything.

Plutus Tx needs to be kept simple and avoid advanced Haskell features. I have never used `QualifiedDo` myself.

Reply:
I wouldn't mind removing all the definitions that rely on `Haskell.Monad` at all. Particularly since I find them either completely useless (like `sequence_` and `mapM_`) or very rarely useful (like `foldrM` and `foldlM`), so the Plutus Tx Prelude could live without them.

> Plutus Tx needs to be kept simple and avoid advanced Haskell features.

I wouldn't call superficial syntactic extension an advanced feature, but YMMV.

> I have never used `QualifiedDo` myself.

Me neither, but I don't use TH in the middle of a definition too, while in Plutus it's perfectly normal. In my opinion, if you have an entirely separate hierarchy of type classes, you can afford using an entirely separate do-notation as well.

Reply:
Yeah, I do think the current situation is pretty suspect and I think I would somewhat prefer either:
1. Remove all the functions we provide that use Haskell`s `Monad`, note that do-notation works somewhat by coincidence.
2. Provide our own `Monad` and related functions, encourage people to use `RebindableSyntax`/`QualifiedDo` or similar.